### PR TITLE
Add `key` type and duplicates verification with hashing

### LINKED ISSUE #2230: Keys yielded while generating dataset are not being checked
The keys used in the dataset generation script to ensure the same order is generated on every user's end should be checked for their types (i.e either `str` or `int`) as well as whether they are unique or not.
Currently, the keys are not being checked for any of these, as evident from `xnli' dataset generation:
https://github.com/huggingface/datasets/blob/56346791aed417306d054d89bd693d6b7eab17f7/datasets/xnli/xnli.py#L196
Even after having a tuple as key, the dataset is generated without any warning.

Also, as tested in the case of `anli` dataset (I tweeked the dataset script to use `1` as a key for every example):
```
>>> import datasets
>>> nik = datasets.load_dataset('anli')
Downloading and preparing dataset anli/plain_text (download: 17.76 MiB, generated: 73.55 MiB, post-processed: Unknown size, total: 91.31 MiB) to C:\Users\nikhil\.cache\huggingface\datasets\anli\plain_text\0.1.0\43fa2c99c10bf8478f1fa0860f7b122c6b277c4c41306255b7641257cf4e3299...
0 examples [00:00, ? examples/s]1        {'uid': '0fd0abfb-659e-4453-b196-c3a64d2d8267', 'premise': 'The Parma trolleybus system (Italian: "Rete filoviaria di Parma" ) forms part of the public transport network of the city and "comune" of Parma, in the region of Emilia-Romagna, northern Italy. In operation since 1953, the system presently comprises four urban routes.', 'hypothesis': 'The trolleybus system has over 2 urban routes', 'label': 'entailment', 'reason': ''}
2021-04-16 12:38:14.483968: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cudart64_110.dll
1 examples [00:01,  1.87s/ examples]1    {'uid': '7ed72ff4-40b7-4f8a-b1b9-6c612aa62c84', 'premise': 'Alexandra Lendon Bastedo (9 March 1946 – 12 January 2014) was a British actress, best known for her role as secret agent Sharron Macready in the 1968 British espionage/science fiction adventure series "The Champions". She has been cited as a sex symbol of the 1960s and 1970s. Bastedo was a vegetarian and animal welfare advocate.', 'hypothesis': "Sharron Macready was a popular character through the 1980's.", 'label': 'neutral', 'reason': ''}
1        {'uid': '5d2930a3-62ac-485d-94d7-4e36cbbcd7b5', 'premise': 'Alexandra Lendon Bastedo (9 March 1946 – 12 January 2014) was a British actress, best known for her role as secret agent Sharron Macready in the 1968 British espionage/science fiction adventure series "The Champions". She has been cited as a sex symbol of the 1960s and 1970s. Bastedo was a vegetarian and animal welfare advocate.', 'hypothesis': "Bastedo didn't keep any pets because of her views on animal rights.", 'label': 'neutral', 'reason': ''}
1        {'uid': '324db753-ddc9-4a85-a825-f09e2e5aebdd', 'premise': 'Alexandra Lendon Bastedo (9 March 1946 – 12 January 2014) was a British actress, best known for her role as secret agent Sharron Macready in the 1968 British espionage/science fiction adventure series "The Champions". She has been cited as a sex symbol of the 1960s and 1970s. Bastedo was a vegetarian and animal welfare advocate.', 'hypothesis': 'Alexandra Bastedo was named by her mother.', 'label': 'neutral', 'reason': ''}
1        {'uid': '4874f429-da0e-406a-90c7-22240ff3ddf8', 'premise': 'Alexandra Lendon Bastedo (9 March 1946 – 12 January 2014) was a British actress, best known for her role as secret agent Sharron Macready in the 1968 British espionage/science fiction adventure series "The Champions". She has been cited as a sex symbol of the 1960s and 1970s. Bastedo was a vegetarian and animal welfare advocate.', 'hypothesis': 'Bastedo cared for all the animals that inhabit the earth.', 'label': 'neutral', 'reason': ''}
```
Here also, the dataset was generated successfuly even hough it had same keys without any warning.

The reason appears to stem from here:
https://github.com/huggingface/datasets/blob/56346791aed417306d054d89bd693d6b7eab17f7/src/datasets/builder.py#L988
Here, although it has access to every key, but it is not being checked and the example is written directly:
https://github.com/huggingface/datasets/blob/56346791aed417306d054d89bd693d6b7eab17f7/src/datasets/builder.py#L992

I would like to take this issue if you allow me. Thank You!

### PR DESCRIPTION
**Pull request by NikhilBartwal**
Closes #2230 
There is currently no verification for the data type and the uniqueness of the keys yielded by the `dataset_builder`.
This PR is currently a work in progress with the following goals:

-  [x] Adding `hash_salt` to `ArrowWriter` so that the keys belonging to different splits have different hash
-  [x] Add `key` arrtibute to `ArrowWriter.write()` for hashing
-  [x] Add a hashing class which takes an input key of certain type (`str`/`int`/anything convertible to string) and produces a 128-bit hash using `hashlib.md5`
-  [x] Creating a function giving a custom error message when non-unique keys are found 
   **[This will take care of type-checking for keys]**
-  [x] Checking for duplicate keys in `writer.write()` for each batch

[**NOTE**: This PR is currently concerned with `GeneratorBasedBuilder` only, for simplification. A subsequent PR will be made in future for `ArrowBasedBuilder`]

@lhoestq Thank you for the feedback. It would be great to have your guidance on this!

### CODE REVIEW

**Comment by NikhilBartwal** [comment]
@lhoestq The tests for key type and duplicate keys have been added and verified successfully.
After generating with an intentionally faulty `mnist` script, when there is an incompatible key type, it shows:

```
Downloading and preparing dataset mnist/mnist (download: 11.06 MiB, generated: 60.62 MiB, post-processed: Unknown size, total: 71.67 MiB) to C:\Users\nikhil\.cache\huggingface\datasets\mnist\mnist\1.0.0\5064c25e57a1678f700d2dc798ffe8a6d519405cca7d33670fffda477857a994...
0 examples [00:00, ? examples/s]2021-04-26 02:50:03.703836: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cudart64_110.dll

FAILURE TO GENERATE DATASET: Invalid key type detected
Found Key [0, 0] of type <class 'list'>
Keys should be either str, int or bytes type
```

In the case of duplicate keys, it now gives:

```
Downloading and preparing dataset mnist/mnist (download: 11.06 MiB, generated: 60.62 MiB, post-processed: Unknown size, total: 71.67 MiB) to C:\Users\nikhil\.cache\huggingface\datasets\mnist\mnist\1.0.0\5064c25e57a1678f700d2dc798ffe8a6d519405cca7d33670fffda477857a994...
0 examples [00:00, ? examples/s]2021-04-26 02:53:13.498579: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cudart64_110.dll
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "f:\datasets\datasets-1\src\datasets\load.py", line 746, in load_dataset
    builder_instance.download_and_prepare(
  File "f:\datasets\datasets-1\src\datasets\builder.py", line 587, in download_and_prepare
    self._download_and_prepare(
  File "f:\datasets\datasets-1\src\datasets\builder.py", line 665, in _download_and_prepare
    self._prepare_split(split_generator, **prepare_split_kwargs)
  File "f:\datasets\datasets-1\src\datasets\builder.py", line 1002, in _prepare_split
    writer.write(example, key)
  File "f:\datasets\datasets-1\src\datasets\arrow_writer.py", line 321, in write
    self.check_duplicates()
  File "f:\datasets\datasets-1\src\datasets\arrow_writer.py", line 331, in check_duplicates
    raise DuplicatedKeysError(key)
datasets.keyhash.DuplicatedKeysError: FAILURE TO GENERATE DATASET !
Found duplicate Key: 234467
Keys should be unique and deterministic in nature
```
Please let me know if this is what we wanted to implement. Thanks!

**Comment by lhoestq** [comment]
This looks pretty cool !
We can make focus on the GeneratorBasedBuilder for now yes.

Do you think we could make the ArrowWriter not look for duplicates by default ?
This way we can just enable duplicate detections when instantiating the writer in the GeneratorBasedBuilder for now.

**Comment by NikhilBartwal** [comment]
Thank you @lhoestq



> Do you think we could make the ArrowWriter not look for duplicates by default ?

We can definitely do that by including a `check_duplicates` argument while instantiating `ArrowWriter()`. 

However, since only `GeneratorBasedBuilder` uses the `write()` function (which includes the detection code) and the others like `ArrowBasedBuilder` use `write_table()` which remains as it was (without duplicate detection). I don't think it would be necessary.

Nonetheless, doing this would require just some small changes. Please let me know your thoughts on this. Thanks!

**Comment by lhoestq** [comment]
I like the idea of having the duplicate detection optional for other uses of the ArrowWriter.
This class is the main tool to write python data in arrow format so I'd expect it to be flexible.
That's why I think by default it shouldn't require users to provide keys or do any duplicates detection.

An alternative would be to subclass the writer to include duplicates detection in another class.

Both options are fine for me, let me know what you think !

**Comment by NikhilBartwal** [comment]
> This class is the main tool to write python data in arrow format so I'd expect it to be flexible.
> That's why I think by default it shouldn't require users to provide keys or do any duplicates detection.

Well, that makes sense as the writer can indeed be used for other purposes as well.

> We can definitely do that by including a `check_duplicates` argument while instantiating `ArrowWriter()`.

I think that this would be the simplest and the more efficient option for achieving this as subclassing the writer only for this would lead to unnecessary complexity and code duplication (in case of `writer()`). 

I will be adding the changes soon. Thanks for the feedback @lhoestq!

**Comment by NikhilBartwal** [comment]
@lhoestq I have pushed the final changes just now. 
Now, the keys and duplicate checking will be necessary only when the `ArrowWriter` is initialized with `check_duplicates=True` specifically (in this case, for `GeneratorBasedBuilders`)
Let me know if this is what was required. Thanks!

**Review by lhoestq — COMMENTED** [review]
That's really cool thanks !
Also pinging @albertvillanova for opinions

Let us know if you need help regarding the tests or other CI failures.
Also feel free to add a test in test_arrow_writer.py to make sure it works as expected :)

**Comment by NikhilBartwal** [comment]
@lhoestq Thanks for the feedback! I will be adding the tests for the same very soon. 

However, I'm not sure as to what exactly is causing the `segmentation fault` in the failing CI tests. It seems to be something from `test_concatenation_table_cast` from `test_table.py`, but I'm not sure as to what exactly. Would be great if you could help. Thanks!

**Comment by lhoestq** [comment]
You can merge master into your branch to fix this issue.
Basically pyarrow 4.0.0 has a segfault issue (which has now been resolved on the master branch of pyarrow).
So until 4.0.1 comes out we changed to using `pyarrow<4.0.0` recently.

**Comment by NikhilBartwal** [comment]
@lhoestq Thanks for the help with the CI failures. Apologies for the multiple merge commits. My local repo got messy while merging which led to this.
Will be pushing the commit for the tests soon!

**Comment by NikhilBartwal** [comment]
Hey @lhoestq, I've just added the required tests for checking key duplicates and invalid key data types.
I think we have caught a nice little issue as 27 datasets are currently using non-unique keys (hence, the failing tests). I'd like to make changes to the faulty datasets' scripts. 

However, I was wondering if I should do that in this PR itself or open a new PR as this might get messy in the same PR. Let me know your thoughts on this. Thanks!

**Comment by NikhilBartwal** [comment]
Thanks a lot for the help @lhoestq. Besides merging the new changes, I guess this PR is completed for now :)

**Comment by lhoestq** [comment]
I just merged the PR, feel free to merge `master` into your branch. It should fix most most of the CI issues. If there are some left we can fix them in this PR :)

**Comment by NikhilBartwal** [comment]
@lhoestq Looks like the PR is completed now. Thanks for helping me out so much in this :)

File: tests/test_arrow_writer.py [hunk 1]
```diff
@@ -89,6 +90,45 @@ def test_write(fields, writer_batch_size):
...
+@pytest.mark.parametrize("writer_batch_size", [None, 2, 10])
+@pytest.mark.parametrize(
+    "fields", [None, {"col_1": pa.string(), "col_2": pa.int64()}, {"col_1": pa.string(), "col_2": pa.int32()}]
+)
+def test_duplicate_keys(fields, writer_batch_size):
+    output = pa.BufferOutputStream()
+    schema = pa.schema(fields) if fields else None
+    with ArrowWriter(
+        stream=output,
+        schema=schema,
```
Comment (lhoestq) [inline tests/test_arrow_writer.py:121]: There's no need to parametrize over the fields ;)
```suggestion
def test_duplicate_keys(fields, writer_batch_size):
    output = pa.BufferOutputStream()
    with ArrowWriter(
        stream=output,
```

File: tests/test_arrow_writer.py [hunk 2]
```diff
@@ -89,6 +90,45 @@ def test_write(fields, writer_batch_size):
...
+@pytest.mark.parametrize("writer_batch_size", [None, 1, 10])
+@pytest.mark.parametrize(
+    "fields", [None, {"col_1": pa.string(), "col_2": pa.int64()}, {"col_1": pa.string(), "col_2": pa.int32()}]
+)
+def test_key_datatype(fields, writer_batch_size):
+    output = pa.BufferOutputStream()
+    schema = pa.schema(fields) if fields else None
+    with ArrowWriter(
+        stream=output,
+        schema=schema,
```
Comment (lhoestq) [inline tests/test_arrow_writer.py:102]:
```suggestion
def test_key_datatype(writer_batch_size):
    output = pa.BufferOutputStream()
    with ArrowWriter(
        stream=output,
```

File: tests/test_arrow_writer.py [hunk 3]
```diff
@@ -89,6 +90,45 @@ def test_write(fields, writer_batch_size):
...
+        writer_batch_size=writer_batch_size,
+        hash_salt="split_name",
+        check_duplicates=True,
+    ) as writer:
+        with pytest.raises(InvalidKeyError):
```
Comment (lhoestq) [inline tests/test_arrow_writer.py:107]: Can you also add a test that makes sure that finalizing after using keys doesn't raise errors ?
The only tests involving keys are for checking errors to we're just missing the one that checks the case where evrything is fine :)

File: src/datasets/arrow_writer.py [hunk 4]
```diff
@@ -173,6 +176,17 @@ def __init__(
...
+        if hash_salt is not None:
+            # Create KeyHasher instance using split name as hash salt
+            self._hasher = KeyHasher(hash_salt)
+        else:
+            self._hasher = KeyHasher("")
+
+        if check_duplicates:
+            self._check_duplicates = True
+        else:
+            self._check_duplicates = False
```
Comment (lhoestq) [inline src/datasets/arrow_writer.py:188]:
```suggestion
        self._check_duplicates = check_duplicates
```

File: src/datasets/arrow_writer.py [hunk 5]
```diff
@@ -196,6 +210,7 @@ def __init__(
         self.current_examples: List[Dict[str, Any]] = []
```
Comment (lhoestq) [inline src/datasets/arrow_writer.py:210]:
```suggestion
        self.current_examples: List[Tuple[Dict[str, Any], str]] = []
```

File: src/datasets/arrow_writer.py [hunk 6]
```diff
@@ -293,18 +311,48 @@ def write_rows_on_file(self):
...
+        # Utilize the keys and duplicate checking when `self._check_duplicates` is passed True
+        if self._check_duplicates:
+            # Create unique hash from key and store as (key, example) pairs
+            hash = self._hasher.hash(key)
+            self.current_examples.append((example, hash))
+            # Maintain record of keys and their respective hashes for checking duplicates
+            self.hkey_record.append((hash, key))
+        else:
+            # Store example as a tuple so as to keep the structure of `self.current_examples` uniform
+            self.current_examples.append((example,))
```
Comment (lhoestq) [inline src/datasets/arrow_writer.py:335]: To follow the typing `Tuple[Dict[str, Any], str]`
```suggestion
            self.current_examples.append((example, ""))
```

**Review by lhoestq — COMMENTED** [review]
Just did another review. It's pretty much ready to merge

I just had other comments, sorry:

File: src/datasets/arrow_writer.py (hunk 4 above)
Comment (NikhilBartwal) [reply src/datasets/arrow_writer.py:188]: Gosh, how did I not see this :(

File: src/datasets/arrow_writer.py (hunk 5 above)
Comment (NikhilBartwal) [reply src/datasets/arrow_writer.py:210]: Thanks for noticing this :)

**Comment by NikhilBartwal** [comment]
Hey @lhoestq, I've added the test and corrected the Cl errors as well. Do let me know if this requires any change. Thanks!

File: src/datasets/arrow_writer.py [hunk 7]
```diff
@@ -258,7 +270,10 @@ def write_examples_on_file(self):
         """Write stored examples from the write-pool of examples. It makes a table out of the examples and write it."""
         if not self.current_examples:
             return
-        cols = sorted(self.current_examples[0].keys())
+
+        # Since current_examples will contain (example, key) or (example, ) tuples
```
Comment (lhoestq) [inline src/datasets/arrow_writer.py:274]:
```suggestion
        # Since current_examples contains (example, key) tuples
```

**Review by lhoestq — APPROVED** [review]
LGTM thank you ! :)

**Comment by NikhilBartwal** [comment]
@lhoestq Thank you for the help and feedback. Feels great to contribute!

**Pull request merged**